### PR TITLE
grpc: avoid grpc-status overwrite on sendLocalReply()

### LINKED ISSUE #15156: http direct response: grpc-status is overwritten when content-type is grpc
*Title*: *grpc-status is overwritten when content-type is grpc*

*Description*: when `grpc-status` header is returned through `response_headers_to_add` envoy will overwrite this header in some cases (see below).

*Repro steps*: bug was introduced on envoy 1.16.0 and it was reproduced on 1.16.2 and 1.17.0.

```bash
➜  ~ curl -H "content-type: application/grpc" -v http://localhost:80
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 80 (#0)
> GET / HTTP/1.1
> Host: localhost
> User-Agent: curl/7.64.1
> Accept: */*
> content-type: application/grpc
>
< HTTP/1.1 200 OK
< grpc-status: 2
< content-type: application/grpc
< date: Tue, 23 Feb 2021 18:00:45 GMT
< server: envoy
< content-length: 0
```

**Expected**: returned `grpc-status` should be 3 (see config below).
**Actual**: returned `grpc-status` was be 2.

**Note1**:  the bug does not appear when `content-type: application/grpc` is missing from the request.
**Note2**:  the bug also appeared when `response_headers_to_add` is configured at the routing level (see config below).

*Config*:
```yaml
static_resources:
  listeners:
  - address:
      socket_address:
        address: 0.0.0.0
        port_value: 80
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        typed_config:
          '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: test
          http_filters:
          - name: envoy.router
            typed_config:
              '@type': type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
          route_config:
            response_headers_to_add:
            - append: false
              header:
                key: grpc-status
                value: "3"
            virtual_hosts:
              domains:
              - '*'
              name: test
              routes:
              - match:
                  prefix: /
                direct_response:
                  status: 200

```

*Logs*:
```
[2021-02-23 18:03:59.845][14][debug][conn_handler] [source/server/connection_handler_impl.cc:501] [C2] new connection
[2021-02-23 18:03:59.846][14][debug][http] [source/common/http/conn_manager_impl.cc:254] [C2] new stream
[2021-02-23 18:03:59.847][14][debug][http] [source/common/http/conn_manager_impl.cc:886] [C2][S8918900023782296153] request headers complete (end_stream=true):
':authority', 'localhost'
':path', '/'
':method', 'GET'
'user-agent', 'curl/7.64.1'
'accept', '*/*'
'content-type', 'application/grpc'

[2021-02-23 18:03:59.847][14][debug][http] [source/common/http/filter_manager.cc:755] [C2][S8918900023782296153] request end stream
[2021-02-23 18:03:59.847][14][debug][http] [source/common/http/filter_manager.cc:839] [C2][S8918900023782296153] Sending local reply with details direct_response
[2021-02-23 18:03:59.847][14][debug][http] [source/common/http/conn_manager_impl.cc:1484] [C2][S8918900023782296153] encoding headers via codec (end_stream=true):
':status', '200'
'grpc-status', '2'
'content-type', 'application/grpc'
'date', 'Tue, 23 Feb 2021 18:03:59 GMT'
'server', 'envoy'

[2021-02-23 18:03:59.852][14][debug][connection] [source/common/network/connection_impl.cc:634] [C2] remote close
[2021-02-23 18:03:59.852][14][debug][connection] [source/common/network/connection_impl.cc:241] [C2] closing socket: 0
[2021-02-23 18:03:59.852][14][debug][conn_handler] [source/server/connection_handler_impl.cc:164] [C2] adding to cleanup list
```

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: Close #15156 Avoid grpc-status will be overwrite on sendLocalReply() when it already exists.
Additional Description:
Risk Level: Low
Testing: Unit
Docs Changes: N/A
Release Notes: 
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/common/http/utility_test.cc [hunk 1]
```diff
@@ -633,6 +633,26 @@ TEST(HttpUtility, SendLocalGrpcReply) {
...
+        EXPECT_NE(headers.GrpcStatus(), nullptr);
+        EXPECT_EQ(headers.getGrpcStatusValue(),
+                  std::to_string(enumToInt(Grpc::Status::WellKnownGrpcStatus::InvalidArgument)));
+        EXPECT_NE(headers.GrpcMessage(), nullptr);
+        EXPECT_EQ(headers.getGrpcMessageValue(), "large");
+      }));
+  Utility::sendLocalReply(
+      is_reset, callbacks,
+      Utility::LocalReplyData{true, Http::Code::PayloadTooLarge, "large",
+                              Grpc::Status::WellKnownGrpcStatus::InvalidArgument, false});
```
Comment (alyssawilk) [inline test/common/http/utility_test.cc:653]: Would this test fail prior to this PR?
I think we'd need another one, where the header modifier function adds a status code other than invalid argument to ensure that status code will be preserved, WDYT?

File: source/common/http/utility.cc [hunk 2]
```diff
@@ -547,10 +547,14 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
   if (local_reply_data.is_grpc_) {
     response_headers->setStatus(std::to_string(enumToInt(Code::OK)));
     response_headers->setReferenceContentType(Headers::get().ContentTypeValues.Grpc);
-    response_headers->setGrpcStatus(
-        std::to_string(enumToInt(local_reply_data.grpc_status_
-                                     ? local_reply_data.grpc_status_.value()
-                                     : Grpc::Utility::httpToGrpcStatus(enumToInt(response_code)))));
+
+    if (response_headers->getGrpcStatusValue().empty()) {
```
Comment (alyssawilk) [inline source/common/http/utility.cc:551]: please add the behavioral change to the current release notes.

Do we have any Envoy filters which would have their status code affected as a result?  If not I'm inclined to consider this a minor bug fix and not require runtime guards but ccing @mattklein123 for a second opinion

**Review by alyssawilk — COMMENTED** [review]
Looks like a solid change, thanks for picking it up!

File: source/common/http/utility.cc (hunk 2 above)
Comment (alyssawilk) [reply source/common/http/utility.cc:551]: ping on this one!

Comment (qiwzhang) [reply source/common/http/utility.cc:551]: This should be fine.  It consists with the way of how other fields, e.g. body_text, are handled:  only set a header if  it is empty.    Normally, sending local_reply needs to set necessary fields for the protocol.  but if they are already set, on the corner cases of local_redirect, or header_to_add, then not to set it. 

So this is minor bug fix.

Comment (alyssawilk) [reply source/common/http/utility.cc:551]: Sounds good, please add it to the release notes as a bug fix then!

Comment (Shikugawa) [reply source/common/http/utility.cc:551]: added release note

**Review by alyssawilk — APPROVED** [review]
@mattklein123 for second pass, and thumbs up on lack of guarding (AFIK this doesn't affect existing filters so I don't have concerns)

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**